Sources/BleTransport/Module/BleModule.swift

```diff
@@ -0,0 +1,257 @@
+                        
+                        switch result {
+                        case .success:
+                            self.listeners[characteristicIdentifier] = ({ dataResult in
+                                completion(ReadResult<R>(dataResult: dataResult))
+                            })
+                        case .failure(let error):
+                            completion(.failure(error))
+                        }
+                    }
+                    addOperation(listenOperation)
+                }
+            }
+        }
+}
+
+// MARK: - Disconnect
+extension BleModule {
+    public func disconnect(completion: ((DisconnectionResult) -> Void)? = nil) {
+        guard let connectedPeripheral = connectedPeripheral else { completion?(.failure(BleModuleError.notConnected)); return }
+        let disconnectOperation = Disconnect(peripheral: connectedPeripheral.cbPeripheral, manager: cbCentralManager, callback: completion)
+        addOperation(disconnectOperation)
+    }
+}
+
+extension BleModule: CBCentralManagerDelegate {
+    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
+        delegate.bluetoothAvailable(central.state == .poweredOn)
+        delegate.bluetoothState(central.state)
+    }
+    
+    public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
+        operationsQueue.operationsOfType(Scan.self).first?.discoveredPeripheral(cbPeripheral: peripheral, advertisementData: advertisementData, rssi: RSSI)
+    }
+    
+    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
+        operationsQueue.operationsOfType(Connect.self).first?.didConnectPeripheral()
+    }
+    
+    public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
+        if let error = error {
+            operationsQueue.operationsOfType(Disconnect.self).first?.fail(error)
+            return
+        }
+        let peripheralIdentifier = PeripheralIdentifier(uuid: peripheral.identifier, name: peripheral.name)
+        operationsQueue.operationsOfType(Connect.self).first?.didDisconnectPeripheral()
+        operationsQueue.operationsOfType(Disconnect.self).first?.didDisconnectPeripheral(peripheral: peripheralIdentifier)
+        clearAfterDisconnect(from: peripheralIdentifier)
+    }
+    
+    /**
+     This mostly happens when either the Bluetooth device or the Core Bluetooth stack somehow only partially completes the negotiation of a connection. For simplicity we treat this as a disconnection event, so we can perform all the same clean up logic.
+     */
+    public func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
+        operationsQueue.operationsOfType(Connect.self).first?.didDisconnectPeripheral()
```

Comment:
We should pass the error to `didDisconnectPeripheral` to avoid the less specific `unexpectedDisconnect` error.